src/runtime/task/mod.rs

```diff
@@ -38,6 +41,73 @@ use waker::make_waker;
 
 pub(crate) const DEFAULT_INLINE_TASKS: usize = 16;
 
+/// To make debugging easier, if a task is assigned a `TaskName(s)` Label,
+/// Shuttle will display the String `s` in addition to the `TaskId` in debug output.
+#[derive(Clone, PartialEq, Eq)]
+pub struct TaskName(String);
+
+impl From<String> for TaskName {
+    fn from(s: String) -> Self {
+        Self(s)
+    }
+}
+
+impl From<&str> for TaskName {
+    fn from(s: &str) -> Self {
+        Self(String::from(s))
+    }
+}
+
+impl std::fmt::Debug for TaskName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<TaskName> for String {
+    fn from(task_name: TaskName) -> Self {
+        task_name.0
+    }
+}
+
+impl<'a> From<&'a TaskName> for &'a String {
+    fn from(task_name: &'a TaskName) -> Self {
+        &task_name.0
+    }
+}
+
+/// By default, when a task or thread T is spawned, it inherits all labels from its parent.
+/// It's often useful to modify or add new Labels to T.  One approach is to put label changes
+/// at the beginning of the closure that is passed to `spawn`, but this approach has the drawback
+/// that the changes are applied only when T is first selected for execution, and the closure
+/// is invoked.  To overcome this drawback, we introduce the `LabelFn` label, which allows
+/// users to define a function that modifies the labels inherited from the parent when T is
+/// initially created.  
+///
+/// # Example
+/// The following example shows how a `LabelFn` can be used to set up names for the next child(ren)
+/// that will be spawned by a parent task.
+/// ```
+/// # use shuttle::current::{me, set_label_for_task, get_name_for_task, LabelFn, TaskName};
+/// # use std::sync::Arc;
+/// // In the parent, set up a `LabelFn` that assigns a name to the child task
+/// shuttle::check_dfs(|| {
+///   set_label_for_task(me(), LabelFn(Arc::new(|labels| { labels.insert(TaskName::from("ChildTask")); })));
+///   shuttle::thread::spawn(|| {
+///     assert_eq!(get_name_for_task(me()).unwrap(), TaskName::from("ChildTask")); // child task already has the name
+///     // ... rest of child
+///   }).join().unwrap();
+/// }, None);
+/// ```
+#[derive(Clone)]
+pub struct LabelFn(pub Arc<dyn Fn(&mut Labels) + Send + Sync + 'static>);
```

Comment:
Instead of this, maybe we could add a new method to `std::thread::Builder` that sets the labels for the new task atomically with its creation? Oh, but there's no builders for async tasks. That's too bad.

Reply:
Yeah, exactly.  It's also easier to add this to existing code that typically just calls `thread::spawn(move || { .. })` or `tokio::spawn(async move { .. })`.